### PR TITLE
Switch to new emitter code

### LINKED ISSUE #127: Switch to new emitter code
We have two different versions of Emitter which is confusing. Switch to the new version and update all examples and code in the documentation to use the new emitters api.

### LINKED ISSUE #144: Duplicate example/test/doc code
We have 3 similar use cases where we want to define common optical setups, such as a Cooke Triplet:

1. The OpticSim.Examples module, for new users to play with in the REPL after installing OpticSim
2. Our docs website https://microsoft.github.io/OpticSim.jl/dev/examples/
3. The TestData module, which is used to run tests and benchmarks

Similarly, common geometrical setups, such as Cuboid, Rectangle and hemisphere are scattered around the source code.

We should try to find a way to group these. For the optical setups, I'm currently trying to consolidate them within OpticSim.Examples, but there are slight differences between the three definitions that I'll have to check with you guys.

### PR DESCRIPTION
**Pull request by alfredclwong**
[Examples preview](https://microsoft.github.io/OpticSim.jl/previews/PR140/examples/)

## Changes
- Update to v0.5.0-DEV (pending detector changes for full v0.5.0 release)
- Remove old Emitter.jl and emitters.md (fixes #127)
  - Rename emitters_new.md to emitters.md
  - Retain RayGenerator types in new file RayGenerator.jl
- Rework examples to use new emitters
  - Split Examples.jl into docs.jl and other.jl to clearly see which examples are used in the docs
  - Use CodeTracking macro @code_string to avoid duplicated code between docs and examples (fixes #144)
- Update tests to use new emitters
- Update Vis to use new emitters
  - Update raygenerator default arg
  - Remove origingen-related draw functions
- Add Maybe and MaybeVector types to OpticSim
- Add a more convenient constructor to Directions.Constant

### CODE REVIEW

File: src/Optical/Emitters/Directions.jl
Comment (BrianGun) [inline src/Optical/Emitters/Directions.jl:34]: @alfredclwong does this cause allocations? It shouldn't but in the past using ... in functions has consistently caused allocations.

**Comment by BrianGun** [comment]
@alfredclwong this looks great. Is it ready to go from draft to regular PR? I'm seeing changes you have made that I have also made on another branch and would like to merge your PR into main as soon as possible so I can sync with your version.

**Comment by alfredclwong** [comment]
> @alfredclwong this looks great. Is it ready to go from draft to regular PR? I'm seeing changes you have made that I have also made on another branch and would like to merge your PR into main as soon as possible so I can sync with your version.

Thanks! It's pretty much ready to go, I just wanted to have a crack at syntax highlighting and to get some feedback on the new docs using `@code_string`. What are the duplicate changes? I agree with the sentiment - merges can be painful.

**Comment by friggog** [comment]
Yeah looks great @alfredclwong! Syntax highlighting would be nice as you say, but already this makes life a lot easier - let's make sure to remove all commented code before merging

File: src/Vis/Visualization.jl
Comment (friggog) [inline src/Vis/Visualization.jl:565]: In general this was because you could position each generator without positioning the sub parts (e.g. move a display to (0,2,3) without having to start by putting the pixel at (0,2,3) - this meant drawing rays became complicated as they had to be transformed for each wrapping class. This also meant splitting out drawing of the actual source panel/circle/point `origingen` just returns the origin generator of the source so that it can drawn directly (without any rays)

This was always a huge mess, but at the time we just wanted to get visualisations working without rewriting everything to do with emitters (now done - hooray!) so hopefully a lot of this code can go

**Comment by friggog** [comment]
It would be nice to be able to colour the source surfaces, and perhaps have the default be something different from the ordinary default surface colour so they are distinct? Can't quite follow where/how these are actually getting drawn now with all the diffs and comments

File: docs/make.jl
Comment (friggog) [inline docs/make.jl:25]: Can probably get rid of the (NEW) now 😃

Comment (alfredclwong) [reply docs/make.jl:25]: I was thinking that we could keep it up for at least one minor version (and maybe even list the old emitters as DEPRECATED) just to make it clear that we switched. But we're probably a small enough project that it doesn't matter. I'll get rid of it.

File: src/Optical/Emitters/Directions.jl
Comment (alfredclwong) [reply src/Optical/Emitters/Directions.jl:34]: We should change it anyway as `direction` has a well defined length. I think I've seen you make this change in another branch, though, so can we hold off doing it in this one?

**Comment by alfredclwong** [comment]
> It would be nice to be able to colour the source surfaces, and perhaps have the default be something different from the ordinary default surface colour so they are distinct? Can't quite follow where/how these are actually getting drawn now with all the diffs and comments

I looked into colouring surfaces a while back because I was working with waveguides which kept being drawn out in very distracting rainbow-like patterns! We had a discussion about how to implement it, and I think Ran wrote some code, but I can't remember it being added to the main project. I'll revisit it later.

File: src/Vis/Visualization.jl
Comment (alfredclwong) [reply src/Vis/Visualization.jl]: deleting it all for now

**Comment by alfredclwong** [comment]
Syntax highlighting is working! See the ```@setup ...``` block in examples.md.

The code still needs reshuffling to follow a single convention (see #144, I agree with Charlie). I'm pretty busy this weekend so it'll be a while until I can sit down and code - should be quite easy work though so anyone can do it if you have the time.

**Comment by BrianGun** [comment]
> > @alfredclwong this looks great. Is it ready to go from draft to regular PR? I'm seeing changes you have made that I have also made on another branch and would like to merge your PR into main as soon as possible so I can sync with your version.
> 
> Thanks! It's pretty much ready to go, I just wanted to have a crack at syntax highlighting and to get some feedback on the new docs using `@code_string`. What are the duplicate changes? I agree with the sentiment - merges can be painful.

The duplicate code (there's not much) is on the optimization branch.

File: src/OpticSim.jl
Comment (alfredclwong) [inline src/OpticSim.jl:21]: is this ok to remove?

File: src/Optical/Emitters/Directions.jl
Comment (alfredclwong) [inline src/Optical/Emitters/Directions.jl:37]: @BrianGun this is copied this from the optimization branch so there shouldn't be any merge conflicts.

File: src/Vis/Visualization.jl
Comment (alfredclwong) [inline src/Vis/Visualization.jl:239]: for convenience (instead of checking `if isnothing(filename)`)

File: src/Vis/Visualization.jl
Comment (alfredclwong) [inline src/Vis/Visualization.jl:592]: is it ok to remove these?

File: test/testsets/OpticalSystem.jl
Comment (alfredclwong) [inline test/testsets/OpticalSystem.jl:9]: does this still fulfil the original purpose of this test?

File: docs/src/emitters.md
Comment (BrianGun) [inline docs/src/emitters.md:86]: Does this new example do the same thing as the old one?

File: src/OpticSim.jl
Comment (BrianGun) [reply src/OpticSim.jl:21]: not used anywhere else in the code so ok to remove.

**Review by BrianGun — APPROVED** [review]

**Pull request merged**
